src/eth/rpc/rpc_subscriptions.rs

```diff
@@ -77,14 +78,53 @@ impl RpcSubscriptions {
                     return Ok(());
                 }
 
+                // store here which subscriptions were cleaned to later log them
+                let mut pending_txs_subs_cleaned = Vec::<RpcClientApp>::new();
+                let mut new_heads_subs_cleaned = Vec::<RpcClientApp>::new();
+                let mut logs_subs_cleaned = Vec::<(RpcClientApp, LogFilterInput)>::new();
+
                 // remove closed subscriptions
-                subs.pending_txs.write().await.retain(|_, s| not(s.sink.is_closed()));
-                subs.new_heads.write().await.retain(|_, s| not(s.sink.is_closed()));
-                subs.logs.write().await.retain(|_, inner_map| {
-                    inner_map.retain(|_, s| not(s.sink.is_closed()));
-                    not(inner_map.is_empty())
+                subs.pending_txs.write().await.retain(|_, sub| {
+                    let should_keep = not(sub.sink.is_closed());
+                    if !should_keep {
+                        pending_txs_subs_cleaned.push(sub.client.clone());
+                    }
+                    should_keep
+                });
+                subs.new_heads.write().await.retain(|_, sub| {
```

Comment:
Consider using `partition_map` from `itertools`.
It will return two Vecs, one with the removed subscriptions and the other with the ones to keep.

Reply:
Cool! Didn't know about this one, only the 'partition' in std::Iterator.

I'll try it out.

Reply:
Tried it and found that the code becomes too verbose and less readable, so I'll merge without this, if you wanna take a shot too tho, I'd say go for it.

(I need to mark as resolved to proceed with merging.)